### DISCUSSION TITLE
아이템 38. 확장할 수 있는 열거 타입이 필요하면 인터페이스를 사용하라

### DISCUSSION THREAD

**Discussion by coalong** [Study-2-Effective-Java/Effective-Java#103] [3. 과제]
# 아이템 38. 확장할 수  있는 열거 타입이 필요하면 인터페이스를 사용하라

## 아이템 설명

> 열거 타입의 사용 이점에 대해서는 앞선 아이템에서 소개했다.
이번 아이템에서는 `열거타입을 사용할 때, 확장이 필요한 경우`에는 어떻게 사용하는지 알아보자.
> 

- 열거 타입이 타입 안전 열거 타입보다 우수하다.
- 단, 타입 안전 열거 패턴은 확장할 수 있으나 열거 타입은 그럴 수 없다는 점이다.
- 아래는 열거 타입과 타입 안전 열거 타입의 예시이다.

### 열거 타입

```java
// 열거 타입은 확장 불가능하다.
enum Operation {
    PLUS, MINUS, TIMES, DIVIDE
}
```

### 타입 안전 열거 타입

```java
public class Operation {
  public static final Operation PLUS = new Operation("+");
  public static final Operation MINUS = new Operation("-");
  public static final Operation TIMES = new Operation("*");
  public static final Operation DIVIDE = new Operation("/");

  private final String symbol;

  protected Operation(String symbol) {
    this.symbol = symbol;
  }
}
```

```java
// 타입 안전 열거 패턴은 확장가능 
public class ExtendedOperation extends Operation {
  public static final ExtendedOperation EXP = new ExtendedOperation("^");
  public static final ExtendedOperation REMAINDER = new ExtendedOperation("%");

 protected ExtendedOperation(String symbol) {
  super(symbol);
  }
}
```

<aside>
💡 사실 대부분 상황에서 열거 타입을 확장하는 건 좋지 않은 생각이다.

</aside>

- Enum 은 상수 집합인데, 확장한 enum 의 요소는 상위 enum 의 요소로 취급하지만, 그 반대는 성립하지 않는 건 이상하다.
- 확장된 타입과 상위 타입들의 모든 상수를 순회할 방법도 마땅치 않다.
- 확장성을 높이려면 고려할 요소가 늘어나 설계와 구현이 더 복잡해진다.

## 확장할 수 있는 열거 타입이 알맞는 상황은?
- 연산 코드 (operation code, 연산 코드의 각 요소는 특정 기계가 수행하는 연산을 뜻한다.)
- 이따금 API 가 제공하는 기본 연산 외에 사용자 확장 연산을 추가할 수 있도록 열어줘야 할 때가 있다.

## 인터페이스를 활용하여 해결!
- 연산 코드용 인터페이스를 정의하고, 그 인터페이스를 구현하는 열거 타입으로 사용한다.
- 아래는 예시 코드이다.

```java
// 연산 코드용 인터페이스를 정의.
public interface Operation {
  double apply(double x, double y);
}
```

```java
// 인터페이스를 구현하는 열거 타입으로 사용.
public enum BasicOperation implements Operaion {
  PLUS("+") {
    public double apply(double x, double y) {return x + y;}
  },
  MINUS("-") {
    public double apply(double x, double y) {return x - y;}
  },
  TIMES("*") {
    public double apply(double x. double y) {return x * y;}
  },
  DIVIDE("/") { 
    public double apply(double x, double y) {return x / y;}
  };
  private final String symbol;

  BasicOperation(String symbol) {
    this.symbol = symbol;
  }

  @Override 
  public String toString() {
    return symbol;
  }
}
```

- 이렇게 하면 Operation 을 구현한 또 다른 열거 타입을 정의하여 BasicOperation 을 대체할 수 있다.
- 아래는 지수 연산(EXP)과 나머지 연산(REMAINDER)을 추가한 예시 코드이다.

```java
// 새로 작성한 연산은 기존 연산을 쓰던 곳이면 어디든 사용 가능. (Operation 인터페이스를 사용하는 곳)
public enum ExtendedOperation implements Operation {
  EXP("^") {
    public double apply(double x, double y) {
      return Math.pow(x, y);
    }
  },
  REMAINDER("%") {
    public double apply(double x, double y) {
      return x % y;
    }
  };

  private final String symbol;

  ExtendedOperation(String symbol) {
    this.symbol = symbol;
  }

  @Override
  public String toString() {
    return symbol;
  }
}
```

- 기존에 Operation 인터페이스를 사용하는 곳에서 기존 연산(BasicOperation)과 확장한 연산(ExtendedOperation)을 모두 사용할 수 있다.
<details>
<summary>이러한 패턴을 사용하는 자바 라이브러리 예시</summary>
<div markdown="1">   

- 자바 라이브러리의 [LinkOption 열거 타입](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/file/LinkOption.html) 은 [CopyOption](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/file/CopyOption.html)과 [OpenOption](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/file/OpenOption.html) 인터페이스를 구현.
  - CopyOption: 파일을 복사하거나 이동하는 방법을 구성하는 개체
  - OpenOption: 파일을 열거나 만드는 방법을 구성하는 개체
  - LinkOption: 심볼릭 링크를 처리하는 방법에 대한 옵션을 정의한다.
    
  ![Untitled](https://user-images.githubusercontent.com/71637466/216815924-51cc691c-706a-4276-99e1-fed2f26937f6.png)

    
  💡 CopyOption 과 OpenOption 인터페이스가 쓰이는 곳에서 LinkOption 의 요소를 사용할 수 있다.

</div>
</details>

    
   

## 타입 수준에서의 확장된 열거 타입

- 개별 인스턴스 수준에서뿐 아니라, **타입 수준에서도 기본 열거 타입 대신 확장된 열거 타입을 사용**할 수 있다.

   
### 타입 수준에서의 확장된 열거 타입 Case 1.

- 첫 번째 방법은 Class 리터럴을 넘겨서 사용하는 것.

```java
public static void main(String[] args) {
  double x = Double.parseDouble(args[0]);
  double y = Double.parseDouble(args[1]);
  test(ExtendedOperation.class, x, y);
}

private static **<T extends Enum<T> & Operation>** void test(**Class<T>** opEnumType, double x, double y) {
  for (Operation op : opEnumType.getEnumConstants())
    System.out.printf("%f %s %f = %f%n", x, op, y, op.apply(x, y));
}
```

- main 메서드는 test 메서드에 ExtendedOperaion의 class 리터럴을 넘겨 확장된 연산들이 무엇인지 알려준다.
    - 여기서 class 리터럴은 한정적 타입 토큰 역할을 한다.
    - **<T extends Enum<T> & Operation> :** Class 객체가 열거 타입인 동시에 Operation 의 하위 타입이어야 한다.
    - 열거 타입이어야 원소를 순회할 수 있고, Operaion 이어야 원소가 뜻하는 연산을 수행할 수 있기 때문이다.

   
   

### 타입 수준에서의 확장된 열거 타입 Case 2.

- 두 번째 방법은 한정적 와일드카드 타입인 Collection<? extends Operation> 을 넘기는 것이다.

```java
public static void main(String[] args) {
  double x = Double.parseDouble(args[0]);
  double y = Double.parseDoulbe(args[1]);
  test(Arrays.asList(ExtendedOperation.values()), x, y);
}

  private static void test(**Collection<? extends Operaion>** operations, double x, double y) {
    for (Operation operation : operations) {
      System.out.printf("%f %s %f = %f%n", x, operation, y, operation.apply(x, y));
    }
  }
}
```

- 이 코드는 덜 복잡하고 test 메서드가 여러 구현 타입의 연산을 조합해 호출할 수 있게 되었다.
- 반면 특정 연산에서는 EnumSet 과 EnumMap 을 사용하지 못한다.

## 인터페이스를 이용해 확장 가능한 열거 타입의 문제점
- 열거 타입끼리는 구현을 상속할 수 없다는 문제가 있다.
- 만약에 아무 상태에도 의존하지 않는 경우, 디폴트 구현을 이용해 인터페이스에 추가하는 방법이 있다.
    - 반면에, 위의 Operation 예는 연산 기호를 저장하고 찾는 로직이 BasicOperation과 ExtendedOperation 모두에 들어가야만 한다.
    - 이 경우에는 중복량이 적으니 문제되지 않지만, 공유하는 기능이 많다면 그 부분을 별도의 도우미 클래스나 정적 도우미 메서드로 분히라는 방식으로 코드 중복을 없앨 수 있다.

## 요약
- 열거 타입 자체는 확장할 수 없지만, 인터페이스와 그 인터페이스를 구현하는 기본 열거 타입을 함께 사용해 같은 효과를 낼 수 있다.
- 이 인터페이스를 구현해 자신만의 열거 타입을 만들 수 있다.
- 그리고 API가 인터페이스 기반으로 작성되었다면 기본 열거 타입의 인스턴스가 쓰이는 모든 곳을 새로 확장한 열거 타입의 인스턴스로 대체해 사용 가능하다.

**Comment by JoisFe** [comment]
인터페이스를 구현한 열거 타입이라... 점점 어질어질 하네요 ㅎㅎ
그래도 타입 안전 열거 타입보다는 코드가 깔끔한 것 같습니다. 
그래도 조금 복잡한 면이 있는 것 같은데요
여기에서 추가적으로 apply 메서드를 재정의하는 부분 또한 #106 에서 람다를 이용한다면 코드가 더 깔끔해져서 가독성이 더 좋아질 것 같습니다. 참고해보세요

정리 감사합니다!!
